bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -5201,7 +5184,31 @@ void ReplicaImp::finishExecutePrePrepareMsg(PrePrepareMsg *ppMsg,
 
   tryToStartOrFinishExecution(false);
 }
-
+void ReplicaImp::handleDbCheckpointCreation(const SeqNum sn, const uint64_t timestamp) {
+  const auto consensus_time_ms = ConsensusTime(timestamp);
+  auto timeSinceLastSnapshot = (std::chrono::duration_cast<std::chrono::seconds>(consensus_time_ms) -
+                                DbCheckpointManager::instance().getLastCheckpointCreationTime())
+                                   .count();
+  const auto &seq_num_to_create_dbcheckpoint = DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot();
+  if ((timeSinceLastSnapshot >= getReplicaConfig().dbSnapshotIntervalSeconds.count() &&
+       sn - DbCheckpointManager::instance().getLastDbCheckpointSeqNum() >= getReplicaConfig().dbCheckPointWindowSize) ||
+      (seq_num_to_create_dbcheckpoint.has_value() && *seq_num_to_create_dbcheckpoint == sn)) {
+    std::optional blockId(DbCheckpointManager::instance().getLastReachableBlock());
+    LOG_INFO(GL,
+             "creating db checkpoint in the next stable sequence number" << KVLOG(timeSinceLastSnapshot, sn, *blockId));
+    DbCheckpointManager::instance().saveDbMetadataToReservedPages(
+        sn, std::chrono::duration_cast<std::chrono::seconds>(consensus_time_ms));
+    // The above should be happen determenistically among all replicas, regardless if they actually create the
+    // checkpoint or not.
+    if (getReplicaConfig().dbCheckpointFeatureEnabled) {
+      DbCheckpointManager::instance().setCheckpointInProcess(true, *blockId);
+      onSeqNumIsStableOneTimeCallbacks_.add(
+          [blockId](SeqNum seqNum) { DbCheckpointManager::instance().createDbCheckpointAsync(seqNum, blockId); });
```

Comment:
I don't fully understand the relationship between the blocked id, the sequence number that is saved in the reserved pages and the sequence number that is passed to the callback on execution.
- the block id is the latest block id at the time of executing this method.
- the sn that is saved in the reserved pages, is also the sn at the time of executing this method and matches the block id.
- the sn that is passed to the CB does not necessarily match the reserved pages sn, no?

Reply:
we cannot "skip" a stable checkpoint, if the replica didn't get to stable checkpoint x, it won't be able to get x + 150

Reply:
Doesn't the window allow 2 checkpoints i.e. it won't be able to get to x+300 ?

Reply:
It is in a size of 300, but 0 indexed - for example, 0 - 299, or 150 - 449